geomesa-spark/geomesa-spark-sql/src/main/scala/org/locationtech/geomesa/spark/GeoMesaSparkSQL.scala

```diff
@@ -273,39 +285,18 @@ object SparkUtils extends LazyLogging {
     new GenericRowWithSchema(res, schema)
   }
 
-  // TODO: optimize so we're not type checking every value
-  def toSparkType(v: Any): AnyRef = v match {
-    case t: Date => new Timestamp(t.getTime)
-    case t       => t.asInstanceOf[AnyRef]
+
+  // Since each attribute's corresponding index in the Row is fixed. Compute the mapping once
+  def getSftRowNameMappings(sft: SimpleFeatureType, sampleRow: Row): List[(String, Int)] = {
+    sft.getAttributeDescriptors.map{ ad =>
+      val name = ad.getLocalName
+      (name, sampleRow.fieldIndex(ad.getLocalName))
+    }.toList
   }
 
-  // JNH: Fix this.  Seriously.
-  def row2Sf(sft: SimpleFeatureType, row: Row, builder: SimpleFeatureBuilder, id: String): SimpleFeature = {
-    import java.{lang => jl}
-    sft.getAttributeDescriptors.foreach {
-      ad =>
-        val name = ad.getLocalName
-        // do we need to do type mapping here?
-        val value = ad.getType.getBinding match {
-          case t if t == classOf[jl.Double]                       => row.getAs[jl.Double](name)
-          case t if t == classOf[jl.Float]                        => row.getAs[jl.Float](name)
-          case t if t == classOf[jl.Integer]                      => row.getAs[jl.Integer](name)
-          case t if t == classOf[jl.String]                       => row.getAs[jl.String](name)
-          case t if t == classOf[jl.Boolean]                      => row.getAs[jl.Boolean](name)
-          case t if t == classOf[jl.Long]                         => row.getAs[jl.Long](name)
-          case t if t == classOf[java.util.Date]                  => row.getAs[java.sql.Timestamp](name) //timestamp extends date
-          case t if t == classOf[com.vividsolutions.jts.geom.Point]            => row.getAs[Point](name)
-          case t if t == classOf[com.vividsolutions.jts.geom.MultiPoint]       => row.getAs[MultiPoint](name)
-          case t if t == classOf[com.vividsolutions.jts.geom.LineString]       => row.getAs[LineString](name)
-          case t if t == classOf[com.vividsolutions.jts.geom.MultiLineString]  => row.getAs[MultiLineString](name)
-          case t if t == classOf[com.vividsolutions.jts.geom.Polygon]          => row.getAs[Polygon](name)
-          case t if t == classOf[com.vividsolutions.jts.geom.MultiPolygon]     => row.getAs[MultiPolygon](name)
-          // JNH: Add Geometry types here.
-
-          case t if      classOf[Geometry].isAssignableFrom(t)    => SQLTypes.GeometryTypeInstance
-          case _                                                  => null
-        }
-        builder.set(name, value)
+  def row2Sf(nameMappings: List[(String, Int)], row: Row, builder: SimpleFeatureBuilder, id: String): SimpleFeature = {
+    nameMappings.foreach{ case (name, index) =>
+      builder.set(name, row.getAs[Object](index))
```

Comment:
Based on some digging and tests I ran, it didn't seem necessary to pattern match and getAs a specific type. Getting the value as Object is sufficient since the SimpleFeatureBuilder will handle conversions.

However, there wasn't a real speed difference between pattern matching and not pattern matching, so if we'd prefer to have it for other reasons I could add it back.